### PR TITLE
perf(megatron): preserve LoRA export dtype

### PR DESCRIPTION
**Pull request by hershg**
## Summary

- Add an opt-in setting that preserves Megatron LoRA tensor dtypes during adapter export.
- Keep FP32 export as the default for backward compatibility.
- Avoid doubling rank-32 BF16 adapter payloads solely for serialization.

Equivalent downstream code reduced the GLM adapter from 61.50 GB to 30.75 GB. GLM 5.2 XID `1049159` completed 10/10 optimizer steps with mean publication falling from 651s to 320s. GLM 5.3 256K XID `1049392` completed four finite optimizer updates and durable checkpoint 4 with native BF16 publication.

This is a performance and capacity change, not evidence that BF16 and FP32 have matched GLM 5.3 learning quality. The setting remains opt-in and FP32 remains the default.

## Testing

```bash
uv run --isolated --extra dev --extra skyrl-train pytest -q tests/backends/skyrl_train/workers/test_lora_export.py
```

All three focused tests and formatting pass. Same-head Train CI reported 1,532 passing tests and one unrelated rate-limiter timing failure (`0.147s < 0.1s`). This fork cannot rerun that workflow without repository-admin permission.

---

> [!NOTE]
> **Low Risk**
> Default export behavior is unchanged; opt-in dtype preservation only affects LoRA weight serialization and on-policy sync payload size/precision.
> 
> **Overview**
> Adds an opt-in **`preserve_export_dtype`** flag on policy LoRA config (`trainer.policy.model.lora.preserve_export_dtype`, default **false**) so Megatron→PEFT adapter export can keep native tensor dtypes instead of always casting to float32.
> 
> Megatron worker LoRA sync now builds adapter state via **`build_lora_adapter_state`**, which clones exported weights under `base_model.model.{name}` and applies the dtype policy. **FP32 export remains the default** for backward compatibility; enabling preservation cuts serialized adapter size and sync time when training in BF16.
> 
> Tests cover CLI typing for the new flag and export naming/dtype/cloning behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3e935056506728c395fc7a959970b0994e5faabd. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: skyrl/backends/skyrl_train/workers/megatron/lora_export.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+def build_lora_adapter_state(
+    adapter_weights: Iterable[tuple[str, torch.Tensor]], *, preserve_dtype: bool
+) -> dict[str, torch.Tensor]:
+    """Clone exported Megatron adapter weights into PEFT's key layout."""
+    adapter_state = {}
+    for name, tensor in adapter_weights:
+        exported_tensor = tensor.clone()
+        if not preserve_dtype:
+            exported_tensor = exported_tensor.float()
```
[the remark this answers is not in the document]
Comment (hershg) [reply skyrl/backends/skyrl_train/workers/megatron/lora_export.py]: Fixed in c1139cdb. The preserve-dtype path still clones, while the legacy FP32 path now casts directly so it keeps non-aliasing semantics without the redundant intermediate allocation. An isolated Torch check verifies both dtypes, values, and distinct storage; format passes.

**Comment by hershg** [comment]
Exact-head CPU CI is clean except for the unrelated timing assertion `tests/train/utils/test_rate_limiter.py::test_initial_burst_allowed`: observed 0.147s versus the test's 0.1s threshold. The same job completed 1,532 tests successfully, and all three focused LoRA-export tests pass. I cannot rerun the upstream workflow from the fork; requesting a maintainer rerun of `skyrl_train_tests`.

**Comment by hershg** [comment]
Follow-up b772b3c9 closes the FP32-source edge case in the earlier allocation cleanup: `Tensor.to(dtype=..., copy=True)` keeps this a single allocation while preserving the legacy guarantee that exported storage never aliases the source, even when the source is already FP32. The regression now covers BF16→FP32, BF16 preservation, and FP32→FP32; all three dtype/value/storage cases pass, and Ruff check/format pass.
